### PR TITLE
Structured logging

### PR DESCRIPTION
**Pull request by jhford**
This PR implements structured logging for the taskcluster-lib-api library as well as adding a requestId to help track specific error messages.

The basic approach is to add two middlewares, then bind a logging instance into the context of the API endpoint handlers.  The first middleware just adds a `uuid.v4()` as the req.requestId property, and sets an "X-Taskcluster-Request-Id" header to this value.  This would allow us to correlate *successful* requests to logs, should we need to.  This is the first middleware to run to ensure that all further middlewares have this requestId.

The second middleware attaches a logger to the request.  This is done early in the middleware chain to ensure that it's available to following middlewares.  Originally, I was creating this logger in the `handler` middleware, but I realised that we'd also want to log things from Auth and Schema middlewares using this.

These loggers have the following log levels, with the interpretation that I've been using so far where we use lib-log:

- Fatal: an error that means the process should crash
- Error: an error which is definitely bad and should probably be logged to sentry as an error
- Warn: important information
- Info: helpful information in production, mainly for after the fact debugging
- Debug: only useful in testing, but useful messages
- Trace: only useful in very specific debugging.

One of the nice properties of lib-log is that by default we *will* log error messages and we need to turn off extraneous ones.  Debug module uses the reverse (except that all our services use `* -this -that*` to get this behaviour.  As well, Debug logs are really messy and hard to follow.

The purpose here is to both log internal middleware outcomes, as well as exposing an easy way for users of taskcluster-lib-api to be able to get useful logs which correlate to actual requests.

By default, the structured logs are a little hard to parse.  We still need to set up log ingestion, but in the meantime, the provided `bunyan` command line utility provides a really nice way to get really easy to read logs.

Here's some sample output from one of the unit tests:
```
13:37:59.173Z TRACE taskcluster-lib-api: calling request handler (apiName=getContext, method=GET, route=/context/, requestId=e1d5ebf2-4476-469c-9141-4b45ccb097a1, url=/v1/context)
13:37:59.173Z TRACE taskcluster-lib-api: hello from the trace logging level (apiName=getContext, method=GET, route=/context/, requestId=e1d5ebf2-4476-469c-9141-4b45ccb097a1, url=/v1/context)
13:37:59.173Z DEBUG taskcluster-lib-api: hello from the debug logging level (apiName=getContext, method=GET, route=/context/, requestId=e1d5ebf2-4476-469c-9141-4b45ccb097a1, url=/v1/context)
13:37:59.173Z  INFO taskcluster-lib-api: hello from the info logging level (apiName=getContext, method=GET, route=/context/, requestId=e1d5ebf2-4476-469c-9141-4b45ccb097a1, url=/v1/context)
13:37:59.173Z  WARN taskcluster-lib-api: hello from the warn logging level (apiName=getContext, method=GET, route=/context/, requestId=e1d5ebf2-4476-469c-9141-4b45ccb097a1, url=/v1/context)
13:37:59.173Z ERROR taskcluster-lib-api: hello from the error logging level (apiName=getContext, method=GET, route=/context/, requestId=e1d5ebf2-4476-469c-9141-4b45ccb097a1, url=/v1/context)
13:37:59.173Z FATAL taskcluster-lib-api: hello from the fatal logging level (apiName=getContext, method=GET, route=/context/, requestId=e1d5ebf2-4476-469c-9141-4b45ccb097a1, url=/v1/context)
13:37:59.173Z ERROR taskcluster-lib-api: stack test (apiName=getContext, method=GET, route=/context/, requestId=e1d5ebf2-4476-469c-9141-4b45ccb097a1, url=/v1/context)
    Error: This is what errors look like
        at Object.<anonymous> (/home/jhford/taskcluster/taskcluster-lib-api/test/context_test.js:30:22)
        at /home/jhford/taskcluster/taskcluster-lib-api/lib/api.js:665:29
        at tryCallOne (/home/jhford/taskcluster/taskcluster-lib-api/node_modules/promise/lib/core.js:37:12)
        at /home/jhford/taskcluster/taskcluster-lib-api/node_modules/promise/lib/core.js:123:15
        at flush (/home/jhford/taskcluster/taskcluster-lib-api/node_modules/asap/raw.js:50:29)
        at _combinedTickCallback (internal/process/next_tick.js:131:7)
        at process._tickDomainCallback (internal/process/next_tick.js:218:9)
13:37:59.175Z TRACE taskcluster-lib-api: finished request handler (apiName=getContext, method=GET, route=/context/, requestId=e1d5ebf2-4476-469c-9141-4b45ccb097a1, url=/v1/context)

```

This was the output from running the following with `export LOG_LEVEL=*:trace` to show *all* messages.

```js
    var api = new subject({
      title:        'Test Api',
      description:  'Another test api',
    });

    api.declare({
      method:   'get',
      route:    '/context/',
      name:     'getContext',
      title:    'Test End-Point',
      description:  'Place we can call to test something',
    }, function(req, res) {
      assert(typeof this.log === 'object');
      for (let lvl of ['trace','debug','info','warn','error','fatal']) {
        assert(typeof this.log[lvl] === 'function');
        this.log[lvl]('hello from the ' + lvl + ' logging level');
      }
      this.log.error(new Error('This is what errors look like'), 'stack test');
      res.status(200).json({myProp: this.myProp});
    });
```

This is the raw output, not passed through bunyan:

```json
{"name":"taskcluster-lib-api","hostname":"jhford-desktop","pid":30670,"apiName":"getContext","method":"GET","route":"/context/","requestId":"cd386073-cbcc-439b-a4ee-f660cb19b447","url":"/v1/context","level":10,"msg":"calling request handler","time":"2018-02-21T13:40:14.078Z","v":0}
{"name":"taskcluster-lib-api","hostname":"jhford-desktop","pid":30670,"apiName":"getContext","method":"GET","route":"/context/","requestId":"cd386073-cbcc-439b-a4ee-f660cb19b447","url":"/v1/context","level":10,"msg":"hello from the trace logging level","time":"2018-02-21T13:40:14.079Z","v":0}
{"name":"taskcluster-lib-api","hostname":"jhford-desktop","pid":30670,"apiName":"getContext","method":"GET","route":"/context/","requestId":"cd386073-cbcc-439b-a4ee-f660cb19b447","url":"/v1/context","level":20,"msg":"hello from the debug logging level","time":"2018-02-21T13:40:14.079Z","v":0}
{"name":"taskcluster-lib-api","hostname":"jhford-desktop","pid":30670,"apiName":"getContext","method":"GET","route":"/context/","requestId":"cd386073-cbcc-439b-a4ee-f660cb19b447","url":"/v1/context","level":30,"msg":"hello from the info logging level","time":"2018-02-21T13:40:14.079Z","v":0}
{"name":"taskcluster-lib-api","hostname":"jhford-desktop","pid":30670,"apiName":"getContext","method":"GET","route":"/context/","requestId":"cd386073-cbcc-439b-a4ee-f660cb19b447","url":"/v1/context","level":40,"msg":"hello from the warn logging level","time":"2018-02-21T13:40:14.079Z","v":0}
{"name":"taskcluster-lib-api","hostname":"jhford-desktop","pid":30670,"apiName":"getContext","method":"GET","route":"/context/","requestId":"cd386073-cbcc-439b-a4ee-f660cb19b447","url":"/v1/context","level":50,"msg":"hello from the error logging level","time":"2018-02-21T13:40:14.079Z","v":0}
{"name":"taskcluster-lib-api","hostname":"jhford-desktop","pid":30670,"apiName":"getContext","method":"GET","route":"/context/","requestId":"cd386073-cbcc-439b-a4ee-f660cb19b447","url":"/v1/context","level":60,"msg":"hello from the fatal logging level","time":"2018-02-21T13:40:14.079Z","v":0}
{"name":"taskcluster-lib-api","hostname":"jhford-desktop","pid":30670,"apiName":"getContext","method":"GET","route":"/context/","requestId":"cd386073-cbcc-439b-a4ee-f660cb19b447","url":"/v1/context","level":50,"err":{"message":"This is what errors look like","name":"Error","stack":"Error: This is what errors look like\n    at Object.<anonymous> (/home/jhford/taskcluster/taskcluster-lib-api/test/context_test.js:30:22)\n    at /home/jhford/taskcluster/taskcluster-lib-api/lib/api.js:665:29\n    at tryCallOne (/home/jhford/taskcluster/taskcluster-lib-api/node_modules/promise/lib/core.js:37:12)\n    at /home/jhford/taskcluster/taskcluster-lib-api/node_modules/promise/lib/core.js:123:15\n    at flush (/home/jhford/taskcluster/taskcluster-lib-api/node_modules/asap/raw.js:50:29)\n    at _combinedTickCallback (internal/process/next_tick.js:131:7)\n    at process._tickDomainCallback (internal/process/next_tick.js:218:9)"},"msg":"stack test","time":"2018-02-21T13:40:14.079Z","v":0}
{"name":"taskcluster-lib-api","hostname":"jhford-desktop","pid":30670,"apiName":"getContext","method":"GET","route":"/context/","requestId":"cd386073-cbcc-439b-a4ee-f660cb19b447","url":"/v1/context","level":10,"msg":"finished request handler","time":"2018-02-21T13:40:14.081Z","v":0}
```

### CODE REVIEW

**Comment by imbstack** [comment]
I am not ignoring this, but have to think on it more. As a first round of review however, I would not like to remove babel compilation or move this from src to lib quite yet if that is ok with you. If we want to do this, we should do it in a separate review and do it to all of our libs at once. I know that consistency is the hobgoblin of little minds, but I have a little mind and it comforts me to do at least some things in just one way.

I think we should talk about the high-level goals of structured logging a bit more as well. I see what you wrote about replacing `debug` in taskcluster/taskcluster-rfcs#35 and I don't completely disagree but I do think perhaps we're being a bit harsh to the humble `debug`. Rather, I think that the point of good structured logging is a stream of defined "events" that have some sort of versioned schema about the payload of an event log line. This allows you to parse the logs in the future with confidence and has a side benefit of documenting what each log line means. I believe that a "structurified debug log" does not provide as many benefits as one might believe. I don't believe this change will actually make it much easier to debug anything via papertrail and that is the only real log thing we have right now. I don't want to have to use bunyan to read logs unless it gives me something really useful in return.

This (and lib-log) are a good start but I want to talk about this more before landing it. Especially we maybe should talk more to the team working on mozilla's unified logging stuff before we settle on an output format etc.

**Comment by djmitche** [comment]
By the way, https://github.com/trentm/node-bunyan#recommendedbest-practice-fields suggests `req_id` as the request ID field

**Comment by djmitche** [comment]
This no longer applies, and anyway has been superseded by other structured logging work.
